crates/services/tx_status_manager/src/manager.rs

```diff
@@ -0,0 +1,339 @@
+    ) {
+        let timestamp = Tai64::now();
+        data.timestamps.entry(timestamp).or_default().push(tx_id);
+        data.statuses.insert(tx_id, tx_status.clone());
+    }
+
+    fn register_status(&self, tx_id: TxId, tx_status: &TransactionStatus) {
+        let mut data = self.data.lock();
+        self.prune_old_statuses(&mut data);
+        self.add_new_status(&mut data, tx_id, tx_status);
+    }
+
+    pub fn status_update(&self, tx_id: TxId, tx_status: TransactionStatus) {
+        tracing::debug!(%tx_id, ?tx_status, "new tx status");
+
+        // TODO[RC]: Capacity checks? - Protected by TxPool capacity checks, except for the squeezed state. Maybe introduce some limit.
+        // TODO[RC]: Purge old statuses? - Remove the status from the manager upon putting the status into storage.
+        // TODO[RC]: Shall we store squeezed out variants as well?
+        self.register_status(tx_id, &tx_status);
+
+        match tx_status {
+            TransactionStatus::Submitted { .. }
+            | TransactionStatus::Success { .. }
+            | TransactionStatus::SqueezedOut { .. }
+            | TransactionStatus::Failure { .. } => (),
+            // TODO[RC]: Handle these new variants
+            TransactionStatus::PreConfirmationSuccess {
+                tx_pointer: _,
+                total_gas: _,
+                total_fee: _,
+                receipts: _,
+                outputs: _,
+            } => todo!(),
+            TransactionStatus::PreConfirmationSqueezedOut { reason: _ } => {
+                // TODO[RC]: Squeezed out should still have the tx_id, not just the reason
+                todo!()
+            }
+            TransactionStatus::PreConfirmationFailure {
+                tx_pointer: _,
+                total_gas: _,
+                total_fee: _,
+                receipts: _,
+                outputs: _,
+                reason: _,
+            } => todo!(),
+        };
+
+        self.tx_status_change
+            .update_sender
+            .send(&TxUpdate::new(tx_id, tx_status.into()));
+    }
+
+    pub fn status(&self, tx_id: &TxId) -> Option<TransactionStatus> {
+        self.data.lock().statuses.get(tx_id).cloned()
+    }
+
+    /// Subscribe to status updates for a transaction.
+    pub fn tx_update_subscribe(&self, tx_id: Bytes32) -> anyhow::Result<TxStatusStream> {
+        self.tx_status_change
+            .update_sender
+            .try_subscribe::<MpscChannel>(tx_id)
+            .ok_or(anyhow!("Maximum number of subscriptions reached"))
+    }
+
+    pub fn notify_skipped_txs(&self, tx_ids_and_reason: Vec<(Bytes32, String)>) {
+        tx_ids_and_reason.into_iter().for_each(|(tx_id, reason)| {
+            let error = Error::SkippedTransaction(reason);
+            self.status_update(
+                tx_id,
+                TransactionStatus::SqueezedOut {
+                    reason: error.to_string(),
+                },
+            );
+        });
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Can we be clearer in these tests under what conditions the different scenarios are met. i.e. put these in terms of "given/when/then". 

`simple_registration` doesn't tell us what is doing the registering, and what is being registered. In the context of writing the code it is obvious, but for someone jumping into the code with no context we can make the tests the place where they gather the context. This test might be better expressed as `status_update__txs_are_present_in_the_manager_statuses`. This begs the question, why do there need to be 4 txs and sleeps? Is this test trying to test multiple things?


This applies to all the tests.

Reply:
https://github.com/FuelLabs/fuel-core/issues/2837